### PR TITLE
feat(scaffolder): allow default kind and namespace args on parseEntityRef filter

### PR DESCRIPTION
**Pull request by rikez**
## Hey, I just made a Pull Request!

Relates to https://github.com/backstage/backstage/issues/19519

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### Template

```yaml
apiVersion: scaffolder.backstage.io/v1beta3
kind: Template
metadata:
  name: parse-entity-ref-filter-test-template
  title: Parse Entity Ref Test
  description: Parse Entity Ref Filter Test Template
  tags:
    - recommended
spec:
  owner: backstage/techdocs-core
  type: documentation
  parameters:
    - title: Fill in some steps
      required:
        - component
      properties:
        component:
          title: Entity Picker
          type: string
          description: Select the component
          ui:field: EntityPicker
          ui:options:
            allowArbitraryValues: false
            catalogFilter:
              kind: Component

  steps:
    - id: fetch
      name: Fetch Component Defailts
      action: catalog:fetch
      input:
        entityRef: ${{ parameters.component }}
        optional: true
    
    - id: log
      name: Parse Entity Ref (Just Ref)
      action: debug:log
      input:
        extra: ${{ steps.fetch.output.entity.spec.owner | parseEntityRef }}

    - id: log-2
      name: Parse Entity Ref With Context
      action: debug:log
      input:
        extra: ${{ steps.fetch.output.entity.spec.owner | parseEntityRef({ defaultKind:"group", defaultNamespace:"another-namespace" }) }}

    - id: log-3
      name: Parse Entity Ref With Context (Kind only)
      action: debug:log
      input:
        extra: ${{ steps.fetch.output.entity.spec.owner | parseEntityRef({ defaultKind:"group" }) }}

    - id: log-4
      name: Parse Entity Ref With Context (Namespace only)
      action: debug:log
      input:
        extra: ${{ steps.fetch.output.entity.spec.owner | parseEntityRef({ defaultNamespace:"another-namespace" }) }}

```

#### Component with `owner: user:guest`

Screenshot 2023-08-28 at 9 32 12 AM
Screenshot 2023-08-28 at 9 32 59 AM
Screenshot 2023-08-28 at 9 33 06 AM
Screenshot 2023-08-28 at 9 33 13 AM

#### Component with `owner: guest`

Screenshot 2023-08-28 at 9 37 30 AM 
Screenshot 2023-08-28 at 9 37 35 AM
Screenshot 2023-08-28 at 9 37 40 AM
Screenshot 2023-08-28 at 9 37 48 AM

ParseEntityRef errors when parsing an entity ref without kind and default kind

```plain
2023-08-28T16:34:25.299Z scaffolder error Failed to parse template string: ${{ steps.fetch.output.entity.spec.owner | parseEntityRef }} with error (unknown path)
  Error: Entity reference "guest" had missing or empty kind (e.g. did not start with "component:" or similar) type=plugin
2023-08-28T16:34:25.320Z scaffolder error Failed to parse template string: ${{ steps.fetch.output.entity.spec.owner | parseEntityRef({ defaultNamespace:"another-namespace" }) }} with error (unknown path)
  Error: Entity reference "guest" had missing or empty kind (e.g. did not start with "component:" or similar) type=plugin
```

### CODE REVIEW

File: plugins/scaffolder-backend/src/lib/templating/filters.ts [hunk 1]
```diff
@@ -27,7 +27,15 @@ export const createDefaultFilters = ({
 }): Record<string, TemplateFilter> => {
   return {
     parseRepoUrl: url => parseRepoUrl(url as string, integrations),
-    parseEntityRef: ref => parseEntityRef(ref as string),
+    parseEntityRef: (
+      ref,
+      defaultKind?: JsonValue,
```
Comment (rikez) [inline plugins/scaffolder-backend/src/lib/templating/filters.ts:32]: Curious if you think there's a better way to provide these optional args. I was looking at [nunjucks filters](https://github.com/mozilla/nunjucks/blob/master/nunjucks/src/filters.js#L395), and couldn't find a different example.

Comment (benjdlambert) [reply plugins/scaffolder-backend/src/lib/templating/filters.ts:32]: Is it possible to mirror the api for `parseEntityRef` itself here?

`{{ parameters.entityRef | parseEntityRef({ defaultKind: 'asd' })` ?

Comment (rikez) [reply plugins/scaffolder-backend/src/lib/templating/filters.ts]: The following works:
```TypeScript
parseEntityRef: (ref, context?) => parseEntityRef(ref as string, context as JsonObject)
```


`{{ parameters.entityRef | parseEntityRef({ defaultKind:'asd' })`

Comment (benjdlambert) [reply plugins/scaffolder-backend/src/lib/templating/filters.ts]: Yeah I think this is something similar to what we're looking for here.

File: plugins/scaffolder-backend/src/lib/templating/filters.ts [hunk 2]
```diff
@@ -27,7 +27,8 @@ export const createDefaultFilters = ({
 }): Record<string, TemplateFilter> => {
   return {
     parseRepoUrl: url => parseRepoUrl(url as string, integrations),
-    parseEntityRef: ref => parseEntityRef(ref as string),
+    parseEntityRef: (ref, context?) =>
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/lib/templating/filters.ts:30]:
```suggestion
    parseEntityRef: (ref: JsonValue, context?: JsonValue) =>
```

File: .changeset/tame-jokes-do.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
+---
+'@backstage/plugin-scaffolder-backend': minor
+---
+
+Improved the `parseEntityRef` Scaffolder filter by introducing the ability for users to provide default kind and/or namespace values. The filter now takes
```
Comment (benjdlambert) [inline .changeset/tame-jokes-do.md:5]: Think that we can simplify this changeset a little bit, but add the examples etc here to it's on documentation somewhere.

Maybe adding a `Built in Filters` section to this part? https://backstage.io/docs/features/software-templates/writing-templates

**Review by benjdlambert — CHANGES_REQUESTED** [review]
Looking pretty good :tada: just a few suggestions!

File: .changeset/tame-jokes-do.md (hunk 3 above)
Comment (rikez) [reply .changeset/tame-jokes-do.md:5]: Added docs for the others filters too, let me know if that's ok.

**Comment by rikez** [comment]
👀

File: docs/features/software-templates/writing-templates.md [hunk 4]
```diff
@@ -601,3 +601,98 @@ output things. You can grab that output using `steps.$stepId.output.$property`.
...
+
+```yaml
+- id: log
+  name: Parse Repo URL
+  action: debug:log
+  input:
+    extra: ${{ parameters.repoUrl | parseRepoUrl }}
+```
+
+- **Input**: `https://github.com/backstage/backstage`
```
Comment (benjdlambert) [inline docs/features/software-templates/writing-templates.md:630]:
```suggestion
- **Input**: `github.com?repo=backstage&org=backstage`
```

File: docs/features/software-templates/writing-templates.md [hunk 5]
```diff
@@ -601,3 +601,98 @@ output things. You can grab that output using `steps.$stepId.output.$property`.
...
+
+```yaml
+- id: log
+  name: Pick
+  action: debug:log
+  input:
+    extra: ${{ parameters.owner | parseEntityRef | pick('name') }}
+```
+
+- **Input**: `group:techdocs`
```
Comment (benjdlambert) [inline docs/features/software-templates/writing-templates.md:680]:
```suggestion
- **Input**: `{ kind: :'Component', namespace: 'default', name: 'techdocs'`
```

File: docs/features/software-templates/writing-templates.md [hunk 6]
```diff
@@ -601,3 +601,98 @@ output things. You can grab that output using `steps.$stepId.output.$property`.
...
+- **Output**: `techndocs`
```
Comment (benjdlambert) [inline docs/features/software-templates/writing-templates.md:681]:
```suggestion
- **Output**: `techdocs`
```

File: docs/features/software-templates/writing-templates.md (hunk 5 above)
Comment (benjdlambert) [reply docs/features/software-templates/writing-templates.md:680]: Could also just make this input `JSONObject` and then return value as `JSONValue`

File: docs/features/software-templates/writing-templates.md [hunk 7]
```diff
@@ -601,3 +601,98 @@ output things. You can grab that output using `steps.$stepId.output.$property`.
...
+
+```yaml
+- id: log
+  name: Project Slug
+  action: debug:log
+  input:
+    extra: ${{ parameters.repoUrl | projectSlug }}
+```
+
+- **Input**: `https://github.com/backstage/backstage`
```
Comment (benjdlambert) [inline docs/features/software-templates/writing-templates.md:697]: This should be the same format as above in previous comment: `github.com?repo=backstage&org=backstage`

**Review by benjdlambert — CHANGES_REQUESTED** [review]
Thanks for this! Sorry for the delay, i've been out on holiday - but just a few comments!

**Review by benjdlambert — APPROVED** [review]
Nice :tada: let's go with this!

**Pull request merged**